rpc/src/optimistically_confirmed_bank_tracker.rs

```diff
@@ -101,31 +104,74 @@ impl OptimisticallyConfirmedBankTracker {
+                    bank.slot()
+                );
+                subscriptions.notify_gossip_subscribers(bank.slot());
+            } else if bank.slot() > bank_forks.read().unwrap().root_bank().slot() {
+                pending_optimistically_confirmed_banks.insert(bank.slot());
+                debug!("notify_or_defer defer notifying for slot {:?}", bank.slot());
+            }
+        }
+    }
+
     pub fn process_notification(
         notification: BankNotification,
         bank_forks: &Arc<RwLock<BankForks>>,
         optimistically_confirmed_bank: &Arc<RwLock<OptimisticallyConfirmedBank>>,
         subscriptions: &Arc<RpcSubscriptions>,
-        pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+        mut pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+        last_notified_slot: &mut Slot,
     ) {
         debug!("received bank notification: {:?}", notification);
         match notification {
             BankNotification::OptimisticallyConfirmed(slot) => {
-                if let Some(bank) = bank_forks
-                    .read()
-                    .unwrap()
-                    .get(slot)
-                    .filter(|b| b.is_frozen())
-                {
+                if let Some(bank) = bank_forks.read().unwrap().get(slot) {
                     let mut w_optimistically_confirmed_bank =
                         optimistically_confirmed_bank.write().unwrap();
-                    if bank.slot() > w_optimistically_confirmed_bank.bank.slot() {
+
+                    if bank.slot() > w_optimistically_confirmed_bank.bank.slot() && bank.is_frozen()
+                    {
                         w_optimistically_confirmed_bank.bank = bank.clone();
-                        subscriptions.notify_gossip_subscribers(slot);
+                    }
+
+                    if slot > *last_notified_slot {
+                        for parent in bank.parents().iter().rev() {
+                            if parent.slot() > *last_notified_slot {
+                                debug!("Calling notify_or_defer for parent {:?}", parent.slot());
+                                Self::notify_or_defer(
+                                    subscriptions,
+                                    bank_forks,
+                                    parent,
+                                    *last_notified_slot,
+                                    &mut pending_optimistically_confirmed_banks,
+                                );
+                            }
+                        }
+                        Self::notify_or_defer(
+                            subscriptions,
+                            bank_forks,
+                            bank,
+                            *last_notified_slot,
+                            &mut pending_optimistically_confirmed_banks,
+                        );
+
+                        *last_notified_slot = slot;
```

Comment:
How about whenever `subscriptions.notify_gossip_subscribers` is called, `last_notified_slot` is checked and updated accordingly? This would mean that `last_notified_slot` is updated below when pending opt confirmed banks are notified rather when they are put in the queue. Functionally the same result but I think it would make the code more resilient and readable

Reply:
I think it would be different. We actually want to update it either the slot is notified via `subscriptions.notify_gossip_subscribers` or put in the pending queue to prevent it from being reprocessed.

Reply:
But I agree the name last_notified_slot is a little misleading as it might not be really notified. It is just processed.

Reply:
Changing to `last_observed_slot` would be fine too

Reply:
I think we do not want to unnecessarily put them into the queue. I will rename it to highest_confirmed_slot -- it is what it really is.